### PR TITLE
Add VerifyUtilityAnalyzer to VerifierBuilder

### LINKED ISSUE #5186: Refactor Verifier to builder approach
There are too many overloads in `Verifier` with too many arguments and duplications. We should refactor it to a builder approach instead. 

The exact approach and API shape will be discussed.

### Original proposal by @Corniel in #4971

The SonarAnalyzer TestFramework is awesome. It contains a lot of options to pick when verifying an analyzer.

However, given the enormous number of overloads makes it a bit clumsy. By providing a builder usage could improve:

``` C#
[TestMethod]
public void BreakOutsideSwitch() =>
    new BreakOutsideSwitch().AddSource(@"TestCases\BreakOutsideSwitch.cs").Verify();
```

And a more complex example:

``` C#
[TestMethod]
public void DeclareTypesInNamespaces_CS_Before8() =>
    new CS.DeclareTypesInNamespaces()
    .AddSource(@"TestCases\DeclareTypesInNamespaces.BeforeCSharp8.cs")
    .WithOptions(ParseOptionsHelper.BeforeCSharp8)
    .Verify();
```

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #5186

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Verifier.cs [hunk 1]
```diff
@@ -88,19 +136,66 @@ public Verifier(VerifierBuilder builder)
...
+        public void VerifyUtilityAnalyzerProducesEmptyProtobuf()     // This should never has any arguments
+        {
+            foreach (var compilation in Compile(false))
+            {
+                DiagnosticVerifier.Verify(compilation, analyzers.Single(), CompilationErrorBehavior.Default);
+                new FileInfo(builder.ProtobufPath).Length.Should().Be(0, "protobuf file should be empty");
+            }
+        }
+
+        public void VerifyUtilityAnalyzer<TMessage>(Action<IReadOnlyList<TMessage>> verifyProtobuf)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Verifier.cs:164]: This breaks the `no arguments` pattern :/ but other ways are more complicated and would move this verifier functionality out of this class.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Verifier.cs [hunk 2]
```diff
@@ -112,8 +123,37 @@ public Verifier(VerifierBuilder builder)
...
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Verifier.cs:135]: This seems an over-complicated way to check the protobuf file contents. I suggest replacing this with a method that returns the content and lets the tests make the checks directly.

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/DummyUtilityAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        private readonly IMessage message;
+
+        protected DummyUtilityAnalyzer(string protobufPath, IMessage message) : base("SDummyUtility", "Dummy title")
+        {
+            this.protobufPath = protobufPath;
+            this.message = message;
+        }
+
+        protected sealed override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterCompilationAction(c =>
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/DummyUtilityAnalyzer.cs:54]: Using `_` in this case will create a parameter named `_` and not a discard :/

[the remark this answers is not in the document]
Comment (costin-zaharia-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/DummyUtilityAnalyzer.cs:54]: It's still a common pattern to name parameters `_` if they are not used

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierTest.cs [hunk 4]
```diff
@@ -432,6 +443,73 @@ public void VerifyCodeFix_NotFixed_VB()
...
+            wasInvoked.Should().BeTrue();
+        }
+
+        [TestMethod]
+        public void VerifyUtilityAnalyzer_UnexpectedProtobuf_CS()
+        {
+            var protobuf = TestHelper.TestPath(TestContext, "Empty.pb");
+            new VerifierBuilder().AddAnalyzer(() => new DummyUtilityAnalyzerCS(protobuf, null)).AddSnippet("// Nothing to see here").WithProtobufPath(protobuf)
+                .Invoking(x => x.VerifyUtilityAnalyzer<LogInfo>(x => x.Should().NotBeEmpty()))
+                .Should().Throw<AssertFailedException>().WithMessage("Expected x not to be empty.");
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierTest.cs:501]: The thing here is that the 1st `Should()` is not assertion, but scaffolding that simulates normal code. And the 2nd `Should()` is actually the only real assertion. That happens to use the same library. I'll do some fake exception there.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/DummyUtilityAnalyzer.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/DummyUtilityAnalyzer.cs:54]: That would be unneeded obfuscation

[the remark this answers is not in the document]
Comment (costin-zaharia-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/DummyUtilityAnalyzer.cs:54]: It's not obfuscation, it's just making obvious the fact that that parameter is not used

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierTest.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (costin-zaharia-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierTest.cs:501]: 1. Do we need this test? It looks to me that it checks if the exception remains unhandled when thrown by the action passed as a parameter.
2. The changes were done only for VB test. The C# one remained unchanged.

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierTest.cs:501]: Yes we do, it proves that the lambda is actually used. And that it should propagate the effect of some `.Should().SomethingWrongHappen()`.

I'll fix the VB part

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Verifier.cs (hunk 2 above)
Comment (costin-zaharia-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Verifier.cs:135]: I think that in the consumers we will replace code like:
```
            new VerifierBuilder().AddAnalyzer(() => new DummyUtilityAnalyzerCS(protobufPath, message)).AddSnippet("// Nothing to see here").WithProtobufPath(protobufPath)
                .VerifyUtilityAnalyzer<LogInfo>(x =>
                {
                    x.Should().ContainSingle().Which.Text.Should().Be("Lorem Ipsum");
                    wasInvoked = true;
                });
            wasInvoked.Should().BeTrue();
```
with
```
            new VerifierBuilder().AddAnalyzer(() => new DummyUtilityAnalyzerCS(protobufPath, message)).AddSnippet("// Nothing to see here").WithProtobufPath(protobufPath)
                .VerifyUtilityAnalyzer<LogInfo>()
                .Should().ContainSingle().Which.Text.Should().Be("Lorem Ipsum");
```

Comment (costin-zaharia-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Verifier.cs:135]: Names can be retouched a bit too. 

I think the end result would be much easier to read and maintain.

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]
Would be great to simplify a bit `VerifyUtilityAnalyzer`

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierTest.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierTest.cs]: UT renamed as discussed offline

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Verifier.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Verifier.cs:137]: Discussed offline

**Review by costin-zaharia-sonarsource — APPROVED** [review]

**Pull request merged**